explainers/signed-exchange-subresource-subtitution-explainer.md

```diff
@@ -0,0 +1,455 @@
+
+[`<link rel="alternate" type="application/signed-exchange" href=...>`](https://html.spec.whatwg.org/multipage/links.html#rel-alternate)
+and the equivalent `Link` header are already defined to declare that the
+referenced document is a reformulation of the current document as a signed
+exchange. To offer signed exchanges for subresources, we propose to use the
+[`anchor` parameter](https://tools.ietf.org/html/rfc8288#section-3.2) to
+identify the replaceable subresource. This may be the first use of the `anchor`
+parameter in the web platform.
+
+
+## Use Cases
+
+### Privacy-preserving prefetching subresources
+
+A user is browsing a news feed site (https://feed.example/). The user clicks a
+link to an article in another site (https://publisher.example/article.html). The
+article is using a large JS library (https://cdn.publisher.example/lib.js) and
+the user must wait for the JS file to be fetched from the server to read the
+article.
+
+If the feed site knows that the article is depending on the JS file, the site
+can insert a link element (`<link rel="prefetch"
+href="https://cdn.publisher.example/lib.js" as="script">`) to prefetch the JS
+file while the user is browsing the feed site. But this leaks that the user's
+feed pages include a page using this JS library before the user has
+consented by clicking on the link.
+
+There is an ongoing
+[spec discussion](https://github.com/w3c/resource-hints/issues/82) to make
+cross-origin navigation prefetch work with double key caching which is proposed
+to protect against
+[HTTP Cache Cross-Site Leaks](http://sirdarckcat.blogspot.com/2019/03/http-cache-cross-site-leaks.html). But as long as the UA needs to fetch the
+resource from the publisher's server, this
+[privacy leak problem](https://wicg.github.io/webpackage/draft-yasskin-wpack-use-cases.html#private-prefetch)
+ can't be solved.
+
+Signed Exchange solve this privacy-preserving prefetching problem for main
+resources. If the publisher is providing the article in signed exchange format
+(article.html.sxg), the UA can prefetch the signed exchange from the feed site’s
+own server while the user is browsing the feed site. But there is no way to
+prefetch subresources in a privacy-preserving manner yet.
+
+Our proposal can solve this problem for subresources:
+1. The publisher provides the script file in signed exchange format
+(lib.js.sxg).
+1. The UA prefetches the signed exchange of the script from the feed site’s own
+server while the user is browsing the feed site.
+1. The HTTP response of the article from the publisher's server
+(https://publisher.example/article.html) has an `allowed-alt-sxg` link header to
+declare exactly what version of the subresource is compatible with this main
+resource.
+   ```
+   Link: <https://cdn.publisher.example/lib.js>;
+         rel="allowed-alt-sxg";
+         header-integrity="sha256-..."
+   ```
+
+
+### Recursive subresource signed exchange prefetch
```

Comment:
I have one question.

I think "allowed-alt-sxg" is useful even when the HTML is not served via signed exchange.

1. An aggregator site prefetches "https://aggregator.example/publisher.example/lib.js.sxg".
2. The user clicks a link to "https://publisher.example/article.html".
3. The HTTP response header has `Link: <https://publisher.example/lib.js>; rel="allowed-alt-sxg"; header-integrity="sha256-..."` header.
4. When the UA detects <script src="./lib.js">, the UA checks the header-integrity attribute value of the link rel="allowed-alt-sxg" header and the header-integrity value of "lib.js.sxg" in cache:
   - If they match, the UA will use "lib.js.sxg" in the cache.
   - Otherwise, the UA will fetch "https://publisher.example/lib.js".

Do you think this use case is not useful?
If so, I will remove this use case and make "allowed-alt-sxg" work only when the HTML is served via signed exchange.

Reply:
I feel like that's unlikely to be helpful: the referrer has to guess that a page with a particular set of preloaded and replaceable subresources will be the next link clicked, without being able to predict exactly which link it is? I do agree that it's technically possible.

Reply:
OK. I agree with you.
I will remove this use case, and merge "Privacy-preserving prefetching subresources" and "Recursive subresource signed exchange prefetch".